crates/bevy_app/src/plugin/mod.rs

```diff
@@ -0,0 +1,175 @@
+///
+/// Plugins can specify a list of other plugins which must be present and will be built before they are.
+/// Every instance of a non-unique dependency will be built before the plugin is built.
+///
+/// Plugins can also suply a list of [`TypeId`] they can substitute for as dependencies. These types do not
+/// necessarily need to be actual plugins.
+///
+/// ## Lifecycle of a plugin
+///
+/// When adding a plugin to an [`App`]:
+/// * the app registers the plugin and waits until the app is started
+/// * once the app is started it will call all registered [`Plugin::build`]
+/// * the dependencies of a plugin will be checked and built before the plugin itself
+/// * the plugin will not be omitted if any other plugin returns it's [`TypeId`] from [`Plugin::subs_for`]
+/// * after building, the [`App`] will wait for all registered [`Plugin::ready`] to return `true`
+/// * it will then call all registered [`Plugin::finish`]
+/// * and finally call all registered [`Plugin::cleanup`]
+pub trait Plugin: Downcast + Any + Send + Sync {
+    /// Configures the [`App`] to which this plugin is added.
+    fn build(&self, app: &mut App);
+
+    /// Has the plugin finished its setup? This can be useful for plugins that need something
+    /// asynchronous to happen before they can finish their setup, like the initialization of a renderer.
+    /// Once the plugin is ready, [`finish`](Plugin::finish) should be called.
+    fn ready(&self, _app: &App) -> bool {
+        true
+    }
+
+    /// Finish adding this plugin to the [`App`], once all plugins registered are ready. This can
+    /// be useful for plugins that depends on another plugin asynchronous setup, like the renderer.
+    fn finish(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Runs after all plugins are built and finished, but before the app schedule is executed.
+    /// This can be useful if you have some resource that other plugins need during their build step,
+    /// but after build you want to remove it and send it to another thread.
+    fn cleanup(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Configure a name which will identify this instance of the [`Plugin`] when debugging.
+    fn name(&self) -> Option<&str> {
+        None
+    }
+
+    /// If the plugin can be meaningfully instantiated several times in an
+    /// [`App`], override this method to return `false`.
+    fn is_unique(&self) -> bool {
+        true
+    }
+
+    /// Returns the [`TypeId`] of other plugins that must be built before the [`Plugin`].
+    ///
+    /// **Note:** These types do not actually need to implement [`Plugin`]. Library authors may wish to
+    /// specify an arbitrary marker type, with the expectation that users will implement their own plugin
+    /// delegate using [`subs_for`](Plugin::subs_for).
+    fn depends_on(&self) -> Vec<TypeId> {
+        Vec::new()
+    }
```

Comment:
What's your reasoning behind using a `Vec` for the return value?

I feel like most dependencies will be known at compile time, so this could be a `&'static [TypeId]` instead.

If you wanted to be able to dynamically specify dependencies, you could take advantage of [Rust 1.75](https://blog.rust-lang.org/2023/12/28/Rust-1.75.0.html#async-fn-and-return-position-impl-trait-in-traits) and use `impl Iterator<Item = TypeId>`.

Reply:
Thanks for the feedback! I'm not sure dependencies can always be known at compile time. Library authors might want to add flags to their plugin struct to enable or disable different functionality, and this might require them to change their dependencies.

Worth noting that I am currently reimplementing this PR from scratch, with a cleaner commit history. The new API I am playing with is:
```rust
fn configure(&self, &mut PluginManifest) -> {
    manifest.add_dependency::<Plugin>(true); // required dependency
    manifest.add_substitution::<Plugin>();
}
```
`PluginManifest` is just a thin wrapper over `HashMap<TypeId, PluginRelation>`. This design new significantly improves error reporting and simplifies logic about non-unique plugins. I believe it may also ease migration when/if we decide to transition plugins into the ECS as entities w/ relations.

Reply:
Interesting! That approach definitely seems a lot easier to use and extend. (It also decreases heap allocations, which is good!) Is your rewrite on a public branch? I'm curious to see what else you changed. :)

Reply:
I appreciate your interest! I have some uncommitted tests right now, but I will publish the new approach tomorrow or monday.

I hope to get the whole thing working later this week. This older approach started from dependencies and worked backwards, but I ran into issues with non-top-level plugin builds. My new approach starts from a generic `PluginSet` type which takes over the `App::add_plugins` api and serves as the canonical representation of the `Plugins<M>` trait:
```rust
let mut first_set = PluginSet::new();
set_ab.add_plugins(PluginA);
let mut second_set = PluginSet::new();
second_set.add_plugins((PluginGroup, (PluginB, PluginC), set_ab));
```
I was able to use this to fix bunch of other issues I was struggling with from in branch, mostly at the cost of changing the `PluginGroup` interface. All I have left is to stitch the two together.